### PR TITLE
Dequeue index & add support for sqlite RETURNING clause

### PR DESCRIPTION
**Pull request by elv-gilles**
* contains #22
* upgrade gorqlite to v0.0.9 & add support for sqlite RETURNING clause
* use RETURNING clause for dequeuing tasks

Performances notes: 
* main improvement comes with addition of the index in #22: 
   *  with 100000 'completed' tasks preloaded, dequeuing took on avg. `91.1ms`without index and `3.9ms' with index 
* when a task can be `dequeue`d, use of the returning clause improves performances (approx twice faster)  
* but degrades (a bit) when there's no task to dequeue.

### CODE REVIEW

**Comment by jslching** [comment]
From Slack:

> i am noticing that something seems to have broken task uniqueness, possibly related to retention? The test I mentioned that is now failing basically fails because the same task runs twice, which revealed a bug in the stats but there shouldnt be two of the same task
> 
> do we check the separate retained task table before enqueueing?

**Comment by elv-gilles** [comment]
@jslching 
> do we check the separate retained task table before enqueueing?

fixed in [b144ab9](https://github.com/qluvio/asynq/pull/24/commits/b144ab924faee1548a095d6b93a70165246539f5)

File: internal/rqlite/rqlite_seedtest.go [hunk 1]
```diff
@@ -30,51 +30,70 @@ func SeedPendingQueue(tb testing.TB, r *RQLite, msgs []*base.TaskMessage, qname
...
 				"WHERE task_uuid=?",
-			state,
+			active,
 			encoded,
 			msg.CompletedAt,
-			expireAt,
+			msg.CompletedAt+int64(statsTTL),
 			retainUntil,
 			msg.ID)
+		//st := Statement(
```
Comment (jslching) [inline internal/rqlite/rqlite_seedtest.go]: Cleanup?

File: internal/rqlite/tasks_dequeue.go [hunk 2]
```diff
@@ -0,0 +1,352 @@
...
+// Returns a dequeueResult instance if a task is found.
+//
+// dequeueMessage uses the sqlite syntax using the RETURNING clause
+// as described here: https://www.sqlite.org/lang_returning.html
+func (conn *Connection) dequeueMessage(now time.Time, serverID string, qname string) (*dequeueResult, error) {
+	op := errors.Op("rqlite.dequeueMessage")
+
+	nowUnix := now.Unix()
+
+	//st := Statement(
```
Comment (jslching) [inline internal/rqlite/tasks_dequeue.go]: Are these still needed?

**Review by jslching — APPROVED** [review]
LGTM

File: internal/sqlite3/encoding/json.go [hunk 3]
```diff
@@ -3,11 +3,18 @@ package encoding
...
+	"errors"
 	"fmt"
 
 	"github.com/hibiken/asynq/internal/sqlite3/command"
 )
 
+var (
+	// ErrTypesColumnsLengthViolation is returned when a results
+	// object doesn't have the same number of types and columns
+	ErrTypesColumnsLengthViolation = errors.New("types and columns are different lengths")
```
Comment (lukseven) [inline internal/sqlite3/encoding/json.go]:
```suggestion
	ErrTypesColumnsMismatch = errors.New("number of types differs from columns")
```

File: internal/sqlite3/encoding/json.go [hunk 4]
```diff
@@ -25,6 +32,56 @@ type Rows struct {
 	Time    float64         `json:"time,omitempty"`
 }
 
+// AssociativeRows represents the outcome of an operation that returns query data.
```
Comment (lukseven) [inline internal/sqlite3/encoding/json.go]: Document the difference to `Rows`

File: internal/sqlite3/statement.go [hunk 5]
```diff
@@ -21,26 +21,31 @@ import (
...
+		Arguments: params,
 	}
 }
 
+func (s *Statement) WithReturning(b bool) *Statement {
+	s.Returning = b
+	return s
+}
+
 // Check returns an error is the count of parameters does not match the count of
```
Comment (lukseven) [inline internal/sqlite3/statement.go]:
```suggestion
// Check returns an error if the count of parameters does not match the count of
```

File: internal/sqlite3/statement.go [hunk 6]
```diff
@@ -21,26 +21,31 @@ import (
 // Statement is structurally identical to *gorqlite.Statement in order to be
 // usable with the gorqlite library as well as the sqlite3 package.
 type Statement struct {
-	Sql        string
-	Parameters []interface{}
-	Warning    string // unused but left for compatibility with *gorqlite.Statement
+	Query     string
+	Arguments []interface{}
+	Returning bool
```
Comment (lukseven) [inline internal/sqlite3/statement.go]: document

**Review by lukseven — APPROVED** [review]

**Pull request merged**